src/Internal/Types/Interval.purs

```diff
@@ -259,35 +266,226 @@ instance
         :+ PNil
     )
 
-instance Show a => Show (Interval a) where
+instance Show a => Show (PlutusInterval a) where
   show = genericShow
 
-instance Ord a => JoinSemilattice (Interval a) where
-  join = hull
+instance (Ord a, Ord a, Ring a) => JoinSemilattice (PlutusInterval a) where
+  join i1 i2 = intervalToPlutusInterval $
+    hull (plutusIntervalToInterval i1) (plutusIntervalToInterval i2)
 
-instance Ord a => BoundedJoinSemilattice (Interval a) where
-  bottom = never
+instance (Ord a, Ring a) => BoundedJoinSemilattice (PlutusInterval a) where
+  bottom = mkPlutusInterval (LowerBound PosInf true) (UpperBound NegInf true)
 
-instance Ord a => MeetSemilattice (Interval a) where
-  meet = intersection
+instance (Ord a, Ring a) => MeetSemilattice (PlutusInterval a) where
+  meet i1 i2 = intervalToPlutusInterval $
+    intersection (plutusIntervalToInterval i1) (plutusIntervalToInterval i2)
 
-instance Ord a => BoundedMeetSemilattice (Interval a) where
-  top = always
+instance (Ord a, Ring a) => BoundedMeetSemilattice (PlutusInterval a) where
+  top = intervalToPlutusInterval always
 
-instance ToData a => ToData (Interval a) where
+instance ToData a => ToData (PlutusInterval a) where
   toData = genericToData
 
-instance FromData a => FromData (Interval a) where
+instance FromData a => FromData (PlutusInterval a) where
   fromData i = genericFromData i
 
-instance EncodeAeson a => EncodeAeson (Interval a) where
-  encodeAeson' (Interval i) = encodeAeson' $ HaskInterval
+instance EncodeAeson a => EncodeAeson (PlutusInterval a) where
+  encodeAeson' (PlutusInterval i) = encodeAeson' $ HaskInterval
     { ivFrom: i.from, ivTo: i.to }
 
-instance DecodeAeson a => DecodeAeson (Interval a) where
+instance DecodeAeson a => DecodeAeson (PlutusInterval a) where
   decodeAeson a = do
     HaskInterval i <- decodeAeson a
-    pure $ Interval { from: i.ivFrom, to: i.ivTo }
+    pure $ PlutusInterval { from: i.ivFrom, to: i.ivTo }
+
+-- | A new abstraction for `Interval` on top of the old one
+-- |(PlutusInterval), it restricts the creation of intervals
+-- |to the only ones that can be used in practice:
+-- | [a,b+1) , [-infinity,b+1), [a,infinity], [-infinity,infinity],
+-- | empty.
+-- see https://github.com/Plutonomicon/cardano-transaction-lib/issues/1041
+data Interval :: Type -> Type
+data Interval a
+  = FiniteInterval a a
+  | LowerRay a
```

Comment:
Let's rename to `StartingAt` and `EndingAt`, or better names of your choice.

Reply:
I forgot to do this before begin to add the `ToData` instance, I will do this change at the end.